.gitmodules

```diff
@@ -22,3 +16,12 @@
 [submodule "third_party/zlib"]
 	path = third_party/zlib
 	url = https://chromium.googlesource.com/chromium/src/third_party/zlib.git
+[submodule "third_party/icu"]
+	path = third_party/icu
+	url = https://github.com/kishiguro/icu.git
+[submodule "build"]
+	path = build
+	url = https://github.com/kishiguro/chromium_build.git
+[submodule "v8"]
+	path = v8
+	url = https://github.com/kishiguro/chromium_v8.git
```

Comment:
The changes to v8 and build seem like they shouldn't be necessary. Can you describe the issue you're running into?

Reply:
v8 changes are for fixing build error on Windows.  Current v8 code uses `icu::` for ICU namespace in `isolate.h` but on Windows, we need to abstract it with `U_ICU_NAMESPACE`.  This error only occurs on Windows not on other platform such as Linux nor macOS.  I'm not sure this is related to ICU Data embedded in library config or not.

``` diff
diff --git a/src/execution/isolate.h b/src/execution/isolate.h
index 3917d8cb4b..b3c13a61db 100644
--- a/src/execution/isolate.h
+++ b/src/execution/isolate.h
@@ -1172,9 +1172,9 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
       kDefaultCollator, kDefaultNumberFormat, kDefaultSimpleDateFormat,
       kDefaultSimpleDateFormatForTime, kDefaultSimpleDateFormatForDate};

-  icu::UMemory* get_cached_icu_object(ICUObjectCacheType cache_type);
+  U_ICU_NAMESPACE::UMemory* get_cached_icu_object(ICUObjectCacheType cache_type);
   void set_icu_object_in_cache(ICUObjectCacheType cache_type,
-                               std::shared_ptr<icu::UMemory> obj);
+                               std::shared_ptr<U_ICU_NAMESPACE::UMemory> obj);
   void clear_cached_icu_object(ICUObjectCacheType cache_type);
   void ClearCachedIcuObjects();

@@ -1691,7 +1691,7 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
       return static_cast<std::size_t>(a);
     }
   };
-  std::unordered_map<ICUObjectCacheType, std::shared_ptr<icu::UMemory>,
+  std::unordered_map<ICUObjectCacheType, std::shared_ptr<U_ICU_NAMESPACE::UMemory>,
                      ICUObjectCacheTypeHash>
       icu_object_cache_;

```

Reply:
build changes are for finding `#include <unicode/uversion.h>` in `icu/source/tools/gencolusb/extract_unsafe_backwards.cpp` and some other ICU library codes.  I agree with this change is not necessary and would be fixed in other part of the code.  I'm not expert of GN/ninjya so I couldn't figure out better solution for it.  It would be nice somebody who understand GN/ninjya suggest right way for the fix.

``` diff
diff --git a/config/gcc/BUILD.gn b/config/gcc/BUILD.gn
index 747245f37..7a74cb1fb 100644
--- a/config/gcc/BUILD.gn
+++ b/config/gcc/BUILD.gn
@@ -38,6 +38,10 @@ config("symbol_visibility_hidden") {
     cflags_cc = [ "-fvisibility-inlines-hidden" ]
     cflags_objcc = cflags_cc
   }
+
+  # Temporally fix of ICU build.
+  cflags += [ "-I../../../../rusty_v8/third_party/icu/source/common" ]
+  cflags += [ "-I../../../../third_party/icu/source/common" ]
 }

 # This config is usually set when :symbol_visibility_hidden is removed.
@@ -45,6 +49,10 @@ config("symbol_visibility_hidden") {
 # which would error out otherwise (e.g. -fsanitize=cfi-unrelated-cast)
 config("symbol_visibility_default") {
   cflags = [ "-fvisibility=default" ]
+
+  # Temporally fix of ICU build.
+  cflags += [ "-I../../../../rusty_v8/third_party/icu/source/common" ]
+  cflags += [ "-I../../../../third_party/icu/source/common" ]
 }

 # The rpath is the dynamic library search path. Setting this config on a link
diff --git a/config/mac/BUILD.gn b/config/mac/BUILD.gn
index de8233bba..4deb03a60 100644
--- a/config/mac/BUILD.gn
+++ b/config/mac/BUILD.gn
@@ -47,6 +47,9 @@ config("compiler") {
   if (export_libcxxabi_from_executables) {
     ldflags += [ "-Wl,-undefined,dynamic_lookup" ]
   }
+
+  cflags += [ "-I../../../../rusty_v8/third_party/icu/source/common" ]
+  cflags += [ "-I../../../../third_party/icu/source/common" ]
 }

 # This is included by reference in the //build/config/compiler:runtime_library
```

Reply:
1. Add a test to tests/test_api.rs demonstrating some i18n JS API
  -> Just added test cases for date and currency format using i18n JS API.
2. Be able to conditionally turn off ICU with a cargo feature flag (Eg cargo build --features=i18n)
  -> Working on this.

Reply:
@kishiguro I think we can help get this working (we are somewhat experts in ninja/gn). I'll get back to you in a few days.